### PR TITLE
Yf stratify chimeric

### PR DESCRIPTION
**Pull request by yfarjoun**
### Description

Adds two stratifiers to CollectSamErrorMetrics that allow exploring the data with respect to Chimeric reads and soft-clipped bases.

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [ ] Added or modified tests to cover changes and any new functionality
- [ ] Edited the README / documentation (if applicable)
- [ ] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [ ] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

### CODE REVIEW

File: src/main/java/picard/sam/SamErrorMetric/ReadBaseStratification.java [hunk 1]
```diff
@@ -646,30 +653,32 @@ public String getSuffix() {
...
+        GC_CONTENT(() -> gcContentStratifier, "The gc content of their read."),
+        READ_ORDINALITY(() -> readOrdinalityStratifier, "The read ordinality (i.e. first or second)."),
+        READ_BASE(() -> currentReadBaseStratifier, "the base in the original reading direction."),
+        READ_DIRECTION(() -> readDirectionStratifier, "The alignment direction of the read (encoded as + or -)."),
+        PAIR_ORIENTATION(() -> readOrientationStratifier, "The reads orientation and ordinality. (into F1R2 or F2R1) Assumes reads are \"innies\"."),
+        PAIR_PROPERNESS(() -> readPairednessStratifier, "The properness of the reads alignment. Looks for indications of chimerism."),
+        REFERENCE_BASE(() -> referenceBaseStratifier, "The reference base in the read's direction."),
+        PRE_DINUC(() -> preDiNucleotideStratifier, "The read base at the previous cycle, and the current reference base."),
+        POST_DINUC(() -> postDiNucleotideStratifier, "The read base at the previous cycle, and the current reference base."),
```
Comment (fleharty) [inline src/main/java/picard/sam/SamErrorMetric/ReadBaseStratification.java]: I think "previous" should be "subsequent"

File: src/main/java/picard/sam/SamErrorMetric/ReadBaseStratification.java [hunk 2]
```diff
@@ -646,30 +653,32 @@ public String getSuffix() {
...
-        NS_IN_READ(() -> nsInReadStratifier, "Stratifies bases according to the number of Ns in the read.");
+        BINNED_HOMOPOLYMER(binnedHomopolymerStratifier::get, "The scale of homopolymer (long or short), the base that the homopolymer is comprised of, and the reference base."),
+        FLOWCELL_TILE(() -> flowCellTileStratifier, "The flowcell and tile where the base was read (taken from the read name)."),
+        READ_GROUP(() -> readgroupStratifier, "The read-group id of the read."),
+        CYCLE(() -> baseCycleStratifier, "The machine cycle during which the base was read."),
+        BINNED_CYCLE(() -> binnedReadCycleStratifier, "The binned machine cycle. Similar to CYCLE, but binned into 5 evenly spaced ranges across the size of the read.  This stratifier may produce confusing results when used on datasets with variable sized reads."),
+        SOFT_CLIPS(() -> softClipsLengthStratifier, "The number of softclipped bases their read has."),
```
Comment (fleharty) [inline src/main/java/picard/sam/SamErrorMetric/ReadBaseStratification.java]: "their" -> "the"

File: src/main/java/picard/sam/SamErrorMetric/ReadBaseStratification.java [hunk 3]
```diff
@@ -705,6 +714,48 @@ public static ReadOrdinality of(final SAMRecord sam) {
         }
     }
 
+    /**
+     * An enum to hold information about the "properness" of a read pair
+     */
+    public enum ProperPaired {
```
Comment (fleharty) [inline src/main/java/picard/sam/SamErrorMetric/ReadBaseStratification.java]: Could you add a test to test this new functionality?

File: src/main/java/picard/sam/AbstractAlignmentMerger.java [hunk 4]
```diff
@@ -699,7 +699,7 @@ private void transferAlignmentInfoToFragment(final SAMRecord unaligned, final SA
      * @param rec SAMRecord whose alignment information will be encoded
      * @return String encoding rec's alignment information according to SA tag in the SAM spec
      */
-    static private String encodeMappingInformation(SAMRecord rec) {
+    static public String encodeMappingInformation(SAMRecord rec) {
```
Comment (fleharty) [inline src/main/java/picard/sam/AbstractAlignmentMerger.java]: final SAMRecord rec

File: src/main/java/picard/sam/SamErrorMetric/ReadBaseStratification.java [hunk 5]
```diff
@@ -653,15 +653,15 @@ public String getSuffix() {
      */
     enum Stratifier implements CommandLineParser.ClpEnum {
         ALL(() -> nonStratifier, "Puts all bases in the same stratum."),
-        GC_CONTENT(() -> gcContentStratifier, "The gc content of their read."),
+        GC_CONTENT(() -> gcContentStratifier, "The GC-content of the read."),
         READ_ORDINALITY(() -> readOrdinalityStratifier, "The read ordinality (i.e. first or second)."),
         READ_BASE(() -> currentReadBaseStratifier, "the base in the original reading direction."),
         READ_DIRECTION(() -> readDirectionStratifier, "The alignment direction of the read (encoded as + or -)."),
-        PAIR_ORIENTATION(() -> readOrientationStratifier, "The reads orientation and ordinality. (into F1R2 or F2R1) Assumes reads are \"innies\"."),
+        PAIR_ORIENTATION(() -> readOrientationStratifier, "The read-pair's orientation and ordinality (encoded as F1R2 or F2R1). Assumes reads are \"innies\"."),
```
Comment (fleharty) [inline src/main/java/picard/sam/SamErrorMetric/ReadBaseStratification.java]: Is this comment true?  Won't it also use F1F2 and R1R2?  Does it really assume reads are innies?

Comment (yfarjoun) [reply src/main/java/picard/sam/SamErrorMetric/ReadBaseStratification.java]: I guess not innies, per se, but it finds the ordinality and strandedness of the current read and predicts the other. I guess that could be fixed...but that's what it currently does.

Comment (yfarjoun) [reply src/main/java/picard/sam/SamErrorMetric/ReadBaseStratification.java]: actually. I take that back. you are right.

**Pull request merged**